tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts

```diff
@@ -1109,5 +1132,171 @@ describe('scalar card', () => {
         },
       });
     }));
+
+    describe('tooltip', () => {
+      function buildTooltipDatum(
+        metadata?: ScalarCardSeriesMetadata,
+        point: Partial<ScalarCardPoint> = {}
+      ): TooltipDatum<ScalarCardSeriesMetadata, ScalarCardPoint> {
+        return {
+          id: metadata?.id ?? 'a',
+          metadata: {
+            type: SeriesType.ORIGINAL,
+            id: 'a',
+            displayName: 'A name',
+            visible: true,
+            color: '#f00',
+            ...metadata,
+          },
+          closestPointIndex: 0,
+          point: {x: 0, y: 0, value: 0, step: 0, wallTime: 0, ...point},
+        };
+      }
+
+      function setTooltipData(
+        fixture: ComponentFixture<ScalarCardContainer>,
+        tooltipData: TooltipDatum[]
+      ) {
+        const lineChart = fixture.debugElement.query(Selector.GPU_LINE_CHART);
+
+        lineChart.componentInstance.tooltipData = tooltipData;
+        lineChart.componentInstance.changeDetectorRef.markForCheck();
+      }
+
+      it('renders the tooltip using the custom template (no smooth)', fakeAsync(() => {
+        store.overrideSelector(selectors.getMetricsScalarSmoothing, 0);
+        const fixture = createComponent('card1');
+        setTooltipData(fixture, [
+          buildTooltipDatum(
+            {
+              id: 'row1',
+              type: SeriesType.ORIGINAL,
+              displayName: 'Row 1',
+              visible: true,
+              color: '#00f',
+            },
+            {
+              x: 10,
+              step: 10,
+              y: 1000,
+              value: 1000,
+              wallTime: new Date('2020-01-01').getTime(),
+            }
+          ),
+          buildTooltipDatum(
+            {
+              id: 'row2',
+              type: SeriesType.ORIGINAL,
+              displayName: 'Row 2',
+              visible: true,
+              color: '#0f0',
+            },
+            {
+              x: 1000,
+              step: 1000,
+              y: -1000,
+              value: -1000,
+              wallTime: new Date('2020-12-31').getTime(),
+            }
+          ),
+        ]);
+        fixture.detectChanges();
+
+        const headerCols = fixture.debugElement.queryAll(
+          Selector.TOOLTIP_HEADER_COLUMN
+        );
+        const headerText = headerCols.map(
+          (col) => col.nativeElement.textContent
+        );
+        expect(headerText).toEqual(['', 'Run', 'Value', 'Step', 'Time']);
+
+        const rows = fixture.debugElement.queryAll(Selector.TOOLTIP_ROW);
+        const tableContent = rows.map((row) => {
+          return row
+            .queryAll(By.css('td'))
+            .map((td) => td.nativeElement.textContent);
+        });
+
+        expect(tableContent).toEqual([
+          ['', 'Row 1', '1000', '10', '1/1/20, 12:00 AM'],
+          // Print the step with comma for readability. The value is yet optimize for
```

Comment:
I think both these comments would be more appropriate near the actual code in scalar_card_component.ng.html